discovery.md

```diff
@@ -0,0 +1,378 @@
+producers. These take two forms:
+
+1. Middleware
+2. Aggregator (or *Events Broker*)
+
+In the Middleware case, discovery is a combined set of producers for where the
+discovery is passed through largely unmodified and subscriber URIs are generally
+absolute (back to the producer).
+
+In the aggregator use case, the aggregator is expected to modify the output of
+discovery and subscription URIs will be relative to the aggregator itself.
+
+## Notations and Terminology
+
+### Notational Conventions
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
+"SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be
+interpreted as described in [RFC 2119](https://tools.ietf.org/html/rfc2119).
+
+### Terminology
+
+This specification defines the following terms:
+
+### Producer
+
+The "producer" is a specific instance, process or device that creates the data
+structure describing the CloudEvent.
+
+### Intermediary
+
+An "intermediary" receives a message containing an event for the purpose of
+forwarding it to the next receiver, which might be another intermediary or a
+consumer. A typical task for an intermediary is to route the event to receivers
+based on the information in the event context.
+
+### Middleware
+
+Specific to discovery and subscription capabilities. Events middleware systems
+provide discovery services that may be directly passing through discovery from
+producers, or may aggregate or otherwise alter the discovery. This includes
+events brokers that may also be relaying discovery AND subscription information
+further upstream to event producers. Multiple middleware systems may exist
+between the producer and the consumer.
+
+### Consumer
+
+The entity receiving the event - the destination for the events related to a
+Subscription when the events are “pushed” by the Event Producer. In some
+instances this might be the same entity as the Subscriber. It could be the
+eventual event consumer or a pub/sub messaging entity which delivers the event
+to the eventual event consumer.
+
+### Subscription
+
+The request for events from an Event Producer or middleware system.
+
+### Event Subscriber
+
+The entity managing the lifecycle of a Subscription on behalf of an Event
+Consumer. In some instances this might be the same entity as the Event Consumer.
+
+## API Specification
+
+The discovery API takes the declarative API design approach. This document
+structure focuses on discovery where each document has a two part key of
+`producer` and `type`. This takes a denormalized view where the composite key
+of documents is the `producer` and `type`. Producer is defined as an arbitrary
+human readable string and type is the CloudEvent type attribute.
+
+// TODO
+// 1) need to rationalize fanout, broker aggregator scenario
+// 2) Does discovery need to be 2 levels? - producers -> types -> sources
+
+### Producer entity Attributes
```

Comment:
I'm struggling with how this API works, and I admit, it could just be I need to be educated.

Let's take the scenario where the Provider is static w.r.t. the types of resources that can produce events, and the list of event types. For example, let's take an object store, with buckets and objects and CRUD events for each. That's a well defined list.

I kind of feel like there should be two flavors of query APIs:
1 - one that lists the types of resources (buckets and objects) and the types of events for each (CRUD). And in this respect the results from a query to this side of things will generally always return the same data because it's static.
2 - one that lists the specific instances of each resources - e.g. the list of buckets or objects I can see and subscribe to. And this one, almost, feels out of scope for us because I would imagine that each Provider already offers a way for people to "list their resources". But, in the interest of allowing full interop perhaps we could suggest simple List() type of APIs.

This popped in my head because below it lists things like 'source' as an attribute and (ignoring ACLs that might change which source+event each user can see), I would think the same set of events should be visible on all buckets.  So, if I think about this as a table I would prefer to see:

| provider | source resource | event |
| --- | --- | --- |
| store | bucket | created |
| store | bucket | deleted |
| store | object | created |
| store | object | deleted |

instead of:

| provider | source resource | event |
| --- | --- | --- |
| store | amys-bucket | created |
| store | amys-bucket | deleted |
| store | johns-bucket | created |
| store | johns-bucket | deleted |
... repeat for all buckets ...
| store | amys-bucket/object1 | created |
| store | amys-bucket/object1 | deleted |
... repeat for all objects in all buckets ...

In other words, I would expect one query to give me a (sort of) schema, and then other queries to give me the specific values I can put into that schema info when I create the subscription.

Am I explaining this well enough? :-)

Reply:
Yes, I think I get your point, and I agree. 😄

Reply:
added examples in the doc for my thinking on this